Mathlib/CategoryTheory/MorphismProperty/Lifting.lean

```diff
@@ -0,0 +1,143 @@
+/-
+Copyright (c) 2024 Jack McKoen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jack McKoen
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Limits
+import Mathlib.CategoryTheory.MorphismProperty.Retract
+
+/-!
+# Left and right lifting properties
+
+Given a morphism property `T`, we define the left and right lifting property with respect to `T`.
+
+We show that the left lifting property is stable under retracts, cobase change, and composition,
+with dual statements for the right lifting property.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+namespace MorphismProperty
+
+/-- The left lifting property (llp) with respect to a class of morphisms. -/
+def llp (T : MorphismProperty C) : MorphismProperty C := fun _ _ f ↦
+  ∀ ⦃X Y : C⦄ ⦃g : X ⟶ Y⦄ (_ : T g), HasLiftingProperty f g
+
+/-- The right lifting property (rlp) with respect to a class of morphisms. -/
+def rlp (T : MorphismProperty C) : MorphismProperty C := fun _ _ f ↦
+  ∀ ⦃X Y : C⦄ ⦃g : X ⟶ Y⦄ (_ : T g), HasLiftingProperty g f
+
+open RetractArrow in
+lemma liftOfRetractLeft {X Y Z W Z' W' : C} {f : X ⟶ Y} {g : Z ⟶ W} {g' : Z' ⟶ W'}
+    (h : RetractArrow g' g) (h' : HasLiftingProperty g f) : HasLiftingProperty g' f where
+  sq_hasLift := fun {u v} sq ↦
+    ⟨h.i.right ≫ h'.lift ⟨show (h.r.left ≫ u) ≫ f = g ≫ (h.r.right ≫ v) by simp [sq.w]⟩,
+      by rw [← Category.assoc, ← i_w, Category.assoc, CommSq.fac_left, ← Category.assoc,
+        retract_left, Category.id_comp],
+      by rw [Category.assoc, CommSq.fac_right, ← Category.assoc, retract_right, Category.id_comp]⟩
```

Comment:
I believe this should be the better design:
```suggestion
lemma _root_.CategoryTheory.RetractArrow.leftLiftingProperty
    {X Y Z W Z' W' : C} {g : Z ⟶ W} {g' : Z' ⟶ W'}
    (h : RetractArrow g' g) (f : X ⟶ Y) [HasLiftingProperty g f] : HasLiftingProperty g' f where
  sq_hasLift := fun {u v} sq ↦ by
    have sq' : CommSq (h.r.left ≫ u) g f (h.r.right ≫ v) := by simp [sq.w]
    exact
      ⟨⟨{ l := h.i.right ≫ sq'.lift
          fac_left := by
            simp only [← cancel_epi h.r.left]
            conv_rhs => rw [← sq'.fac_left]
            simp [← h.i_w_assoc]
          fac_right := by simp}⟩⟩
```

Reply:
The general lemmas about `HasLiftingProperty` should go into a file in `Mathlib.CategoryTheory.LiftingProperties`.

Reply:
Thanks for the feedback! What does `_root_` denote?

Reply:
`_root_` refers to the root namespace. It is a way to define a lemma while you are working inside a namespace, but give it a name outside the namespace.